### PR TITLE
Rewrite physics worker in C++ (to strip C# dependency).

### PR DESCRIPTION
**Pull request by dgavedissian**
As a drive by, also enabled PIC compilation so this compiles fine on Ubuntu 18.04

### CODE REVIEW

File: workers/cpp_physics/physics.cc [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+const std::uint32_t kGetOpListTimeoutInMilliseconds = 100;
+
+int main(int argc, char** argv) {
+    auto print_usage = [&]() {
+        std::cout << "Usage: Physics <hostname> <port> <worker_id>" << std::endl;
+        std::cout << std::endl;
+        std::cout << "Connects to SpatialOS" << std::endl;
+        std::cout << "    <hostname>      - hostname of the receptionist to connect to." << std::endl;
+        std::cout << "    <port>          - port to use if connecting through the receptionist." << std::endl;
+        std::cout << "    <worker_id>     - (optional) name of the worker assigned by SpatialOS." << std::endl;
```
Comment (improbable-andreaskrugersen) [inline workers/cpp_physics/physics.cc]: It's not optional since you require 4 parameters below

File: workers/cpp_physics/physics.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    worker::ConnectionParameters parameters;
+    parameters.WorkerType = "physics";
+    parameters.Network.ConnectionType = worker::NetworkConnectionType::kTcp;
+    parameters.Network.UseExternalIp = false;
+    worker::Connection connection = worker::Connection::ConnectAsync(
+        ComponentRegistry{}, std::string{argv[1]}, atoi(argv[2]), argv[3], parameters).Get();
+
+    if (connection.GetConnectionStatusCode() != worker::ConnectionStatusCode::kSuccess) {
+        std::cerr << "Failed to connection the receptionist." << std::endl;
```
Comment (improbable-andreaskrugersen) [inline workers/cpp_physics/physics.cc]: Failed to connect to?

**Review by improbable-andreaskrugersen — APPROVED** [review]

File: workers/cpp_physics/physics.cc (hunk 1 above)
Comment (dgavedissian) [reply workers/cpp_physics/physics.cc]: lol, yes, I forgot to update this :P

File: workers/cpp_physics/physics.cc [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+#include <improbable/worker.h>
+#include <improbable/view.h>
+
+#include <improbable/standard_library.h>
+#include <client_data.h>
+
+using ComponentRegistry = worker::Components<improbable::Position, improbable::Metadata>;
+
+const int kErrorExitStatus = 1;
+const std::uint32_t kGetOpListTimeoutInMilliseconds = 100;
```
Comment (Mudit-G) [inline workers/cpp_physics/physics.cc]: nit - `Get` seems odd if it isn't a function. Maybe just `kOpListTimeoutMs`.

File: workers/cpp_physics/physics.cc [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            is_connected = false;
+        } else {
+            std::cout << "[remote] " << op.Message << std::endl;
+        }
+    });
+    view.OnCommandRequest<sample::Login::Commands::TakeControl>([&](const worker::CommandRequestOp<sample::Login::Commands::TakeControl>& op) {
+        connection.SendLogMessage(worker::LogLevel::kInfo, "Physics", "Assigning ClientData component to " + op.CallerAttributeSet[1]);
+
+        // Assign write ACL of the client data component to the worker ID received.
+        improbable::EntityAcl::Update acl_update;
```
Comment (Mudit-G) [inline workers/cpp_physics/physics.cc:70]: The `EntityAcl` isn't in the component registry at the moment.

File: workers/cpp_physics/physics.cc [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    view.OnLogMessage([&](const worker::LogMessageOp& op) {
+        if (op.Level == worker::LogLevel::kFatal) {
+            std::cerr << "Fatal error: " << op.Message << std::endl;
+            is_connected = false;
+        } else {
+            std::cout << "[remote] " << op.Message << std::endl;
+        }
+    });
+    view.OnCommandRequest<sample::Login::Commands::TakeControl>([&](const worker::CommandRequestOp<sample::Login::Commands::TakeControl>& op) {
+        connection.SendLogMessage(worker::LogLevel::kInfo, "Physics", "Assigning ClientData component to " + op.CallerAttributeSet[1]);
```
Comment (Mudit-G) [inline workers/cpp_physics/physics.cc:67]: `op.CallerAttributeSet[1]`, is it still guaranteed to have two attributes?

File: workers/cpp_physics/physics.cc [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        connection.SendComponentUpdate<improbable::EntityAcl>(entityId, acl_update);
+    });
+    view.OnCommandResponse<sample::ClientData::Commands::TestCommand>([&](const worker::CommandResponseOp<sample::ClientData::Commands::TestCommand>& op) {
+        if (op.StatusCode == worker::StatusCode::kSuccess) {
+            connection.SendLogMessage(worker::LogLevel::kInfo, "Physics", "Received command response: " + std::to_string(op.Response->sum()));
+        }
+    });
+
+    int tick_count = 0;
+	float angle = 0;
```
Comment (Mudit-G) [inline workers/cpp_physics/physics.cc]: weird formatting.

File: workers/cpp_physics/physics.cc [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    while (is_connected) {
+        view.Process(connection.GetOpList(kGetOpListTimeoutInMilliseconds));
+
+        // Update position of entity.
+        angle += 0.5f;
+		improbable::Position::Update position_update;
```
Comment (Mudit-G) [inline workers/cpp_physics/physics.cc]: I can't tell if this is just the Github view, but indentation seems odd here as well.

File: workers/cpp_physics/physics.cc [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		position_update.set_coords({sin(angle) * 10, 0.0, cos(angle) * 10});
+		connection.SendComponentUpdate<improbable::Position>(entityId, position_update);
+
+        // Sleep for some time.
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+
+        // Send a command every 2 ticks.
+        if (tick_count % 2 == 0) {
```
Comment (Mudit-G) [inline workers/cpp_physics/physics.cc]: Could do `if (tick_count++ % 2 == 0)`

**Review by Mudit-G — COMMENTED** [review]
Sorry, I don't really know CMake well enough to give useful review on those changes.

File: README.md [hunk 9]
```diff
@@ -18,14 +18,30 @@ spatial local launch
...
 * Client (vtable): `spatial local worker launch client_vtable local`
 
+No output will be written, but you should see a 2nd worker connect in the inspector. To view the
+printf logging from the C workers, launch the executable directly. For example, on Windows:
+
+```
+./workers/c_client_direct/cmake_build/Debug/ClientDirect.exe localhost 7777 ClientDirect1
+```
+
+(note that the 3rd parameter should be a different string every time you run a worker).
```
Comment (Helcaraxan) [inline README.md]: Nit: make it more explicit that the string should be unique for a single deployment. A deployment restart allows for string reuse.

File: workers/cpp_physics/.gitignore [hunk 10]
```diff
@@ -0,0 +1,9 @@
+# This worker is using a custom build.json
+!build.json
+!spatialos_worker_packages.json
+
+# CMake
+cmake_build/
+
+# Generated code
+generated_code/
```
Comment (Helcaraxan) [inline workers/cpp_physics/.gitignore]: Missing end-of-line

File: workers/cpp_physics/physics.cc (hunk 5 above)
Comment (dgavedissian) [reply workers/cpp_physics/physics.cc:67]: Yes.

File: workers/cpp_physics/physics.cc (hunk 6 above)
Comment (dgavedissian) [reply workers/cpp_physics/physics.cc]: Yeah, this is some tabs that should be spaces.

File: workers/cpp_physics/physics.cc (hunk 4 above)
Comment (dgavedissian) [reply workers/cpp_physics/physics.cc:70]: Yep you're right (and there are also some other missing components).

File: README.md (hunk 9 above)
Comment (dgavedissian) [reply README.md]: I decided to fix this by making the clients generate a random suffix alongside the ID.

**Review by Helcaraxan — APPROVED** [review]

**Pull request merged**